src/Readers/PseudoTopReader.cpp

```diff
@@ -0,0 +1,247 @@
+    newLeptonicW_.reset();
+    newLepton_.reset();
+    newAllLeptons_.clear();
+    newNeutrino_.reset();
+    newJets_.clear();
+    newPseudoBs_.clear();
+    newPseudoMET_.reset();
+    isSemiLeptonic_ = true;
+
+    readPseudoTopParticles();
+    return pseudoTopParticles_;
+}
+
+void PseudoTopReader::readPseudoTopParticles() {
+
+
+    ParticlePointer wPlus( new Particle() );
+    ParticlePointer wMinus( new Particle() );
+
+    // Go through pseudo top decay chain and extract info
+    for (unsigned int index = 0; index < pseudoTop_pdgIdReader.size(); index++) {
+        int pdgId = pseudoTop_pdgIdReader.getIntVariableAt(index);
+        double energy = pseudoTop_energyReader.getVariableAt(index);
+        double px = pseudoTop_pxReader.getVariableAt(index);
+        double py = pseudoTop_pyReader.getVariableAt(index);
+        double pz = pseudoTop_pzReader.getVariableAt(index);
+
+        // Pseudo tops
+        if ( fabs(pdgId) == 6 ) {
+            newPseudoTops_.push_back( ParticlePointer( new Particle( energy, px, py, pz )) );
+        }
+        // Leptons
+        else if ( fabs(pdgId) == 11 || fabs(pdgId) == 13 ) {
+
+            // If there's more than one dressed lepton, this event is not a semi leptonic psuedo ttbar event
+            if ( newLepton_ != 0 ) {
+                isSemiLeptonic_ = false;
+                // Store lepton with highest pt
+                MCParticle anotherNewLepton( energy, px, py, pz);
+                if ( anotherNewLepton.pt() > newLepton_->pt() ) {
+                    newLepton_ = MCParticlePointer( new MCParticle( anotherNewLepton ) );
+                    newLepton_->setPdgId( pdgId );                    
+                }
+
+            }
+            else {
+                newLepton_ = MCParticlePointer( new MCParticle( energy, px, py, pz ) );
+                newLepton_->setPdgId( pdgId );
+            }
+        }
+        // Neutrino
+        else if ( fabs(pdgId) == 12 || fabs(pdgId) == 14 || fabs(pdgId) == 16 ) {
+
+            if ( newNeutrino_ != 0 ) {
+                // Store neutrino with highest pt
+                Particle anotherNewNeutrino( energy, px, py, pz);
+                if ( anotherNewNeutrino.pt() > newNeutrino_->pt() ) {
+                    newNeutrino_ = ParticlePointer( new Particle( anotherNewNeutrino ) );
+                }
+            }
+            else {
+                newNeutrino_ = ParticlePointer( new Particle( energy, px, py, pz));
+            }
+        }
+        // Bs
+        else if ( fabs(pdgId) == 5 ) {
+            newPseudoBs_.push_back( ParticlePointer( new Particle( energy, px, py, pz )) );            
+        }
+        // Ws
+        else if ( fabs( pdgId ) == 24 ) {
+            if ( pdgId > 0 ) {
+                wPlus = ParticlePointer( new Particle( energy, px, py, pz));
+            }
+            else if ( pdgId < 0 ) {
+                wMinus = ParticlePointer( new Particle( energy, px, py, pz));
+            }
```

Comment:
On particle level it is possible to determine the flavour, but only if it has been set. You can also set the charge with ```Particle::setCharge(double)``` and access it via ```Particle::charge()```.